handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -1122,6 +1162,36 @@ private SSLEngineResult wrap(ByteBufAllocator alloc, SSLEngine engine, ByteBuf i
                 in0[0] = newDirectIn.internalNioBuffer(newDirectIn.readerIndex(), readableBytes);
             }
 
+            // Check if the engine type wants a mutable buffer and if so replace all readonly buffers with mutable ones.
+            if (engineType.wantsMutableBuffer) {
+                // count the number of readonly buffers to avoid an extra allocation if no readonly buffers are found.
+                int readOnlyBufferCount = 0;
+                for (int i = 0; i < in0.length; i++) {
+                    if (in0[i].isReadOnly()) {
+                        readOnlyBufferCount++;
+                    }
+                }
+                if (readOnlyBufferCount > 0) {
+                    newMutableBuffers = new ByteBuf[readOnlyBufferCount];
+                    int newMutableBufferIndex = 0;
+                    for (int i = 0; i < in0.length; i++) {
+                        if (in0[i].isReadOnly()) {
+                            ByteBuffer original = in0[i];
+                            ByteBuf copyOfNioBuffer;
+                            if (engineType.wantsDirectBuffer) {
+                                copyOfNioBuffer = alloc.directBuffer(original.remaining());
+                            } else {
+                                copyOfNioBuffer = alloc.buffer(original.remaining());
+                            }
+                            copyOfNioBuffer.writeBytes(original);
+                            in0[i] = copyOfNioBuffer.internalNioBuffer(copyOfNioBuffer.readerIndex(),
+                                                                       copyOfNioBuffer.readableBytes());
+                            newMutableBuffers[newMutableBufferIndex++] = copyOfNioBuffer;
+                        }
+                    }
+                }
+            }
```

Comment:
I didn't yet find a way to add tests for this logic, so that's something TODO. Which test class should the test go in?

Reply:
I guess you could add something to the `SslHandlerTest` and just put a handler in front that will make the buffers readOnly before pass them to the next handler. This way this should be triggered when using an older JDK (which we do in our test matrix).